utils/server_auth.go

```diff
@@ -0,0 +1,342 @@
+	if err != nil {
+		return errors.Wrap(err, "Failed to load director's public JWK")
+	}
+	tok, err := jwt.Parse([]byte(strToken), jwt.WithKey(jwa.ES256, key), jwt.WithValidate(true))
+	if err != nil {
+		return errors.Wrap(err, "Failed to verify JWT by director's key")
+	}
+
+	scopeValidator := createScopeValidator(anyOfTheScopes)
+	if err = jwt.Validate(tok, jwt.WithValidator(scopeValidator)); err != nil {
+		return errors.Wrap(err, "Failed to verify the scope of the token")
+	}
+
+	c.Set("User", "Director")
+	return nil
+}
+
+// Check token authentication with token obtained from authOption.Sources, found the first
+// token available and proceed to check against a list of authOption.Issuers with
+// authOption.Scopes, return true and set "User" context to the issuer if any of the issuer check succeed
+//
+// Scope check will pass if your token has ANY of the scopes in authOption.Scopes
+func CheckAnyAuth(ctx *gin.Context, authOption AuthOption) bool {
+	token := ""
+	errMsg := ""
+	// Find token from the provided sources list, stop when found the first token
+	tokenFound := false
+	for _, opt := range authOption.Sources {
+		if tokenFound {
+			break
+		}
+		switch opt {
+		case Cookie:
+			cookieToken, err := ctx.Cookie("login")
+			if err != nil || cookieToken == "" {
+				errMsg += fmt.Sprintln("No 'login' cookie present: ", err)
+				continue
+			} else {
+				token = cookieToken
+				tokenFound = true
+				break
+			}
+		case Header:
+			headerToken := ctx.Request.Header["Authorization"]
+			if len(headerToken) <= 0 {
+				errMsg += fmt.Sprintln("No Authorization header present")
+				continue
+			} else {
+				token = strings.TrimPrefix(headerToken[0], "Bearer ")
+				tokenFound = true
+				break
+			}
+		case Authz:
+			authzToken := ctx.Request.URL.Query()["authz"]
+			if len(authzToken) <= 0 {
+				errMsg += fmt.Sprintln("No Authz query parameter present")
+				continue
+			} else {
+				token = authzToken[0]
+				tokenFound = true
+				break
+			}
+		default:
+			log.Info("Authentication failed. Invalid/unsupported token source")
+			return false
+		}
+	}
+
+	if token == "" {
+		log.Info("Authentication failed. No token is present from the list of potential token positions")
+		return false
+	}
+
+	for _, iss := range authOption.Issuers {
+		switch iss {
+		case Federation:
+			err := authChecker.FederationCheck(ctx, token, authOption.Scopes)
+			if _, exists := ctx.Get("User"); err != nil || !exists {
+				errMsg += fmt.Sprintln("Federation Check failed: ", err)
+				log.Debug("Federation Check failed: ", err)
+				break
+			} else {
+				log.Debug("Federation Check succeed")
+				return exists
+			}
+		case Director:
+			err := authChecker.DirectorCheck(ctx, token, authOption.Scopes)
+			if _, exists := ctx.Get("User"); err != nil || !exists {
+				errMsg += fmt.Sprintln("Director Check failed: ", err)
+				log.Debug("Director Check failed: ", err)
+				break
+			} else {
+				log.Debug("Director Check succeed")
```

Comment:
So when I use this new code to serve a director, I see the log output `Director Check succeeded`. It's always good to see success, but I'm having trouble understanding what was being checked and what success implies. Can you elaborate a bit?

Reply:
I'd like to elaborate more but I'm not sure if the extra verbosity is what we want for the original ticket #405 where Brian suggests not to have too much verbose message, but I guess for the DEBUG level output it should be fine?

Reply:
It doesn't even have to be at the level of logging output. I'm just curious what this does and how I should interpret it (even if we don't change any code for now).